Force throw of out-of-bound during parsing of arrays

freetom opened the pull request:
Hi,

I'm using binpac to implement a complex parser and embed its functionality with the Bro NIDS. 

During the development and testing of the software I realized that in some particular cases binpac does not throw an out-of-bound exception while parsing an array with invalid size (it should throw the exception). Furthermore, if the array is the last element of the current packet then nothing else will raise an exception. By consequence, binpac will go to the message's linked routine to generate the Bro event and call a routine to convert the binpac vector to a compatible Bro policy script type. My routine will then slide on the array to extract its elements. If, at this point, I use the field used in the formal specifications to specify the length of the array I will get the wrong value and the code will get a segmentation fault because it will try to read more elements than how many are actually in the vector.

The present patch is pretty rough but with it the problem is solved because binpac raises the exception as it's supposed to do. I didn't observed any malfunctioning of binpac with my mods

note: the added throw statement does not specify the violated bounds because I was not able to always put the right variables there, but the comments should help. For you that shouldn't be hard to adjust

I hope you understand the problem I'm pointing at but let me know if I can clarify and/or help.


Signed-off-by: Tomas Bortoli <tomasbortoli@gmail.com>

Commit:
Fixed check to prevent the ignoring of out-of-bound during parsing of arrays

note: the added throw statement does not specify the violated bounds because I was not able to always put the right variables there. For the binpac creators it shouldn't be hard to adjust.

Signed-off-by: Tomas Bortoli <tomasbortoli@gmail.com>

rsmmr commented:
I'm not sure this actually correct. I looked at what this patch changes in terms of code generation and see instances like this example from the DNP3 analyzer:

```
--- build.old/src/analyzer/protocol/dnp3/dnp3_pac.cc    2017-07-07 13:31:08.652806713 -0700
+++ build/src/analyzer/protocol/dnp3/dnp3_pac.cc        2017-07-07 14:11:08.129942400 -0700
@@ -821,7 +821,10 @@
                                                for (; /* forever */; ++t_read_requests__elem__it)
                                                        {
                                                        if ( t_read_requests__elem__dataptr >= t_end_of_data )
                                                                {
                                                                read_requests__elem_ = 0;
-                                                               goto end_of_read_requests;
+                                                               // Handle out-of-bound condition
+                                                               throw binpac::ExceptionOutOfBound("Request_Objects *",
+                                                                       0, //current_data_pointer- t_begin_of_data+1
+                                                                0); //(t_end_of_data) - (t_begin_of_data)
                                                                }
                                                        read_requests__elem_ = new Request_Objects(app_header()->function_code());
                                                        int t_read_requests__elem__size;
```

That if-condition there used to be the normal (and only) way to end the loop, without however being an error condition. With this patch that would now throw an exception. 

There are other locations where there's actually already an out-of-bounds check  after such an if-condition, even without the patch. Maybe the code emitting that existing check could be adjusted to catch more situations, like the one you're running into. (disclaimer: I"m saying that without really knowing binpac internal very well). Here's an example of that from mysql (again using the diff output showing what your patch is adding; but note the existing oob check):

```
@@ -1289,7 +1295,10 @@
                // Check &until(session_id__elem__dataptr >= end_of_data)
                if ( t_session_id__elem__dataptr >= t_end_of_data )
                        {
-                       goto end_of_session_id;
+                       // Handle out-of-bound condition
+                       throw binpac::ExceptionOutOfBound("uint8",
+                               0, //current_data_pointer- t_begin_of_data+1
+                        0); //(t_end_of_data) - (t_begin_of_data)
                        }
                // Checking out-of-bound for "V2ClientHello:session_id__elem"
                if ( t_session_id__elem__dataptr + (1) > t_end_of_data || t_session_id__elem__dataptr + (1) < t_session_id__elem__dataptr )
```

Commit:
Modified ArrayType::GenUntilCheck function to check for out-of-bound
exceptions if the array length is known and thence the check is
possible (plus call SetBoundaryChecked function to evict multiple
out-of-bound check). Otherwise, put in place until check.

freetom commented:
I modified the code so that it checks for out-of-bounds only if the number of array elements is known (that is exactly the buggy case I found). In this way the arrays which length is determined by _&until_ or _restofdata_ are treated by the code as before (with the until check and `goto`).

The two buggy/repetitive checks cases of analyzers you highlighted are now o.k (plus, the second code snippet come from the SSL analyzer, not MYSQL)

I could not found a better way to patch. I unfortunately didn't got yet all the related binpac logic .


Diffs between old and new `*_pac.cc` attached.

[dnp3_diff.txt](https://github.com/bro/binpac/files/1133177/dnp3_diff.txt)
[ssl_diff.txt](https://github.com/bro/binpac/files/1133176/ssl_diff.txt)

Commit:
Turned out that some oob checks by my routine were not proper, in the cases of uint16, uint32 and the other native typed arrays (because I was checking only for one byte after). Therefore this patch just avoid the 'until' check if the number of elements of the array is known and thence oob exception to manage. Otherwise, binpac will behave as previously, putting in place an oob check BUT without the 'until' check that would make binpac skip some obb exceptions in corner cases

freetom commented:
In the end, I realized that to fix the issue, the best solution was to simply remove the 'until' check for arrays with fixed (known) number of elements (so binpac will not skip the corner oob-exceptions but catch them)

The final patch consists in just one more `if` statement before the code that places the 'until' check.

Should be fine now. Looking forward a confirm. Diff of old vs new parsers for SSL and DNP3 attached. 


[ssl-diff.txt](https://github.com/bro/binpac/files/1211134/ssl-diff.txt)
[dnp3-diff.txt](https://github.com/bro/binpac/files/1211136/dnp3-diff.txt)

freetom commented:
@rsmmr, hey, can I have feedback on the patch?

rsmmr commented:
Sorry to be a pain but I'm still not comfortable with this. Here's an example difference from DNP3 between old and new:  
```
                                const_byteptr t_Reserved__elem__dataptr = (t_dataptr_after_signature + 26);
                                for (; t_Reserved__elem__it < t_Reserved__arraylength; ++t_Reserved__elem__it)
                                        {
-                                       // Check &until(Reserved__elem__dataptr >= end_of_data)
-                                       if ( t_Reserved__elem__dataptr >= t_end_of_data )
-                                               {
-                                               goto end_of_Reserved;
-                                               }
                                        Reserved__elem_ = FixByteOrder(t_byteorder, *((uint16 const *) (t_Reserved__elem__dataptr)));
                                        // Evaluate 'let' and 'withinput' fields
                                        Reserved_->push_back(Reserved__elem_);
```

The old version would abort the loop if the data ptr exceeds the available input, which means it doesn't just trust that the given array size (which is constant 4 here) aligns with the data available. The new version would just go ahead and read beyond the end of data pointer in the FixByteOrder() call. Am I missing something?

freetom commented:
I can't get the diff you posted. How did you get it?

The last diff I posted between old and new DNP3 parser  is:
[dnp3-diff.txt](https://github.com/bro/binpac/files/1358392/dnp3-diff.txt)

freetom commented:
I made an example that should clarify a bit.

I attach you a zip that contains an example parser that has only 1 type made of an array of fixed size.
I compiled it with the unpatched and patched Binpac and attached a diff too. You'll see that the patched parser's logic catches the oob exception instead of skipping it, by triggering the condition of the `if` (not present in `parser_new.cc`) and jumping out of the loop.

[ExampleParser.zip](https://github.com/bro/binpac/files/1358799/ExampleParser.zip)


The general idea behind the patch is: that specific `if`, that jumps out of the `for` loop when parsing arrays should only be present when parsing arrays of unspecified size. Because that's the only way to understand they're finished (there's no more data to parse). On the opposite, when an array has fixed size, the Binpac's parser should raise an exception if the length isn't the right one. Therefore it makes sense to remove that `if` and let the exception to be caught by the successive parsing statements.

In the example I made there is an array composed of native elements but even in the case of an array of objects, Binpac will instantiate an object per each round (by calling the constructor) and it'll jump into its parsing routine, that will then raise the exception as expected.



Hope I've been clear enough

rsmmr commented:
I get all of that except for one part: it doesn't seem guaranteed that there's an out-of-bounds check done by the next parse operation. In my example above the call to FixByteOrder() reads directly from  `t_Reserved__elem__dataptr`. The only thing limiting how for that pointer moves inside the loop is the array length inside the `for` condition; and that value can generally not be trusted to be correct. (To answer your question where that example is from: it's not from DNP3, sorry, it's from PE).

How about this: we don't remove the `if` when there's no &until but instead raise an exception right there if the check fails. That goes back to your earlier idea but limits it to cases where we'd otherwise be trusting the array length. Would that work?

freetom commented:
> I get all of that except for one part: it doesn't seem guaranteed that there's an out-of-bounds check done by the next parse operation. In my example above the call to FixByteOrder() reads directly from t_Reserved__elem__dataptr. The only thing limiting how for that pointer moves inside the loop is the array length inside the for condition; and that value can generally not be trusted to be correct. (To answer your question where that example is from: it's not from DNP3, sorry, it's from PE).

What is PE for? I couldn't find the exact code you posted by "grepping" through the whole Bro's codebase.

Nonetheless, I found a very similar example. By typing `grep -i "t_Reserved__elem__data" -rn -A28 -B40` I found two code snippets in the SMB parser that I attach here because I think it's basically the same case/type of code you referred to. 
[snippets.txt](https://github.com/bro/binpac/files/1369070/snippets.txt)



So, in the snippets we've a call to `FixByteOrder` that directly reads from the buffer to the array without apparently checking for oob. Hover, a wider look at the code, shows that the functions at issue, make an oob check at the beginning (because it already knows the size of the structure that it has to parse). Therefore, the oob check is not present in the for cycle because the function already performed the check! Thence I think that an oob error should not happen in the `for` cycles at issue.

Please verify my statements and let me know if there are other cases in which the patch might break the parsers' compilation.




I also observed another buggy behaviour while investigating on the issues here. When generating a parser by specifying a field of an elementary type, with variable length like `x: uint32 &length=8;`.
Binpac will generate code only to parse a word and not 8 words. This is more a problem of specifications because it should raise an exception like "Cannot set &length on someting that is not a bytestring".
The exact same issue arises when doing something like `x: uint32 &restofdata;`. The generated C++ code will only parse a word of 32 bit but not words of 32 bits until the end of the data. These are not security issues as missing it is to miss malformed packets (malformed arrays' checks)
Example attached. 
[issue2.zip](https://github.com/bro/binpac/files/1367993/issue2.zip)

rsmmr commented:
Ok, I think you're right, there are a few variations of the pattern but in the end it looks like there's always a length enforcement before the loop is entered. Now, I've merged this locally, however I'm now seeing test failures, pretty certainly because binpac now raises out-of-bounds exceptions in situations that it accepted earlier. Specifically, two tests from the btest test suite fail (scripts.base.protocols.mysql.wireshark, scripts.base.protocols.smb.raw-ntlm); plus I see some significant differences with SMB traffic on a larger trace I have (which I can send you if you want to investigate). I can't easily tell if these diffs might be benign, but Bro is definitely logging quite a bit information on that larger trace with the patch applied ...

freetom commented:
> Ok, I think you're right, there are a few variations of the pattern but in the end it looks like there's always a length enforcement before the loop is entered. Now, I've merged this locally, however I'm now seeing test failures, pretty certainly because binpac now raises out-of-bounds exceptions in situations that it accepted earlier. Specifically, two tests from the btest test suite fail (scripts.base.protocols.mysql.wireshark, scripts.base.protocols.smb.raw-ntlm); plus I see some significant differences with SMB traffic on a larger trace I have (which I can send you if you want to investigate). I can't easily tell if these diffs might be benign, but Bro is definitely logging quite a bit information on that larger trace with the patch applied ...

So the parsers at issue are counting on the weak-oob checking for arrays of Binpac (voluntarily or not) to successfully parse the relative PDUs.

Issues derived from the Binpac with the current patch can only be linked to the parsing of **fixed-length** arrays as that is the only difference.

The [MYSQL parser](https://github.com/bro/bro/blob/master/src/analyzer/protocol/mysql/mysql-protocol.pac) has few fixed-length arrays, therefore it shouldn't be too hard to find the issue in the source by having knowledge of the protocol's internal mechanisms.

SMB looks much more complicated. But the point is that, if the patch in the present PR is correct, then the buggy behaviour is due to how the parsers are written.

jsiwek commented:
I came back upon this as I was working on https://bro-tracker.atlassian.net/browse/BIT-1829, and looking more into the patch in this PR and the test suite differences, I think the way I've addressed it may be better.

Specifically, the OOB checks that happen at the beginning of parsing functions that were mentioned as part of the SMB snippets.txt aren't always there, and they actually don't always cover the **full** bounds checking for records containing fixed-length arrays.  So I think at least the rational for removing the per-loop-iteration check was based on an incorrect assumption.

The MySQL test case that was failing appears to be one such case where the "until(dataptr >= end_of_data)" check was needed for the array-parsing-loop to terminate at the right point.  And actually, it's an equality-check-to-terminate-loop combined with a throw-when-past-the-end that I think makes sense here so that's what I've changed it to in the topic/jsiwek/bit-1829 branch

I'll leave this PR open for a bit in case there's rebuttal or other opinions, otherwise expect it to be closed sometime after the JIRA ticket gets closed.

freetom commented:
I've merged your branch and compiled a sample parser to see the diffs.

I still think that there's a problem (the one for which I opened the PR btw). The `goto` is still there. If a parser receives an array of fixed size, let's say 100 elements and the array has less than 100 elements and the array is the last element of the packet, I think the parser would get into the `if` of the for cycle, the one with the `goto` and it'd jump out of the loop and return the parsed object (as the pointer to the data would be pointing exactly to the end of the data). Nonetheless, the parser should have thrown an exception to signal that the packet contained an array of wrong length. Correct? I am assuming that binpac should report such malformed packets.

The whole point of this PR was to find a way to enforce the checking of arrays of fixed size (size decided by a constant or variable that can be another field of the packet) to detect whether the array it's really of that size of whether it's not. 

Furthermore, I've compiled the sample parser I attached in a [previous comment](https://github.com/bro/binpac/pull/4#issuecomment-334417078) and I get this for-loop to parse the array:
```
for (; t_array__elem__it < t_array__arraylength; ++t_array__elem__it)
		{
		// Check &until(array__elem__dataptr == end_of_data)
		if ( t_array__elem__dataptr == t_end_of_data )
			{
			goto end_of_array;
			}
		if ( t_array__elem__dataptr > t_end_of_data )
			{
			throw binpac::ExceptionOutOfBound("A:array",
			  (t_array__elem__dataptr) - (t_end_of_data));
			}
		// Checking out-of-bound for "A:array__elem"
		if ( t_array__elem__dataptr + (4) > t_end_of_data || t_array__elem__dataptr + (4) < t_array__elem__dataptr )
			{
			// Handle out-of-bound condition
			throw binpac::ExceptionOutOfBound("A:array__elem",
				((t_array__elem__dataptr - t_begin_of_data)) + (4), 
				(t_end_of_data) - (t_begin_of_data));
			}
		array__elem_ = FixByteOrder(t_byteorder, *((uint32 const *) (t_array__elem__dataptr)));
		// Evaluate 'let' and 'withinput' fields
		array_->push_back(array__elem_);
		t_array__elem__dataptr += 4;
		}
```

The check you added: `if ( t_array__elem__dataptr > t_end_of_data )` is already checked by the oob check that follows it. Maybe there are cases where it's not but in this one it is useless.

jsiwek commented:
Thanks, I do see and agree that the first point still is a problem and the later optimization should also be done, so I'm in the process trying out some changes to address those.  (However, may have arrived back at the point where some protocol parsers in Bro may need to be audited for bugs).

freetom commented:
Thank you for contributing.. So, we'll probably need 2 steps to fix this:

-  Patch binpac to generate parsers that detect arrays with **invalid length**
-  Patch the Bro's parsers that fail with the new behavior

The patch I proposed is minimal and I don't see why it shouldn't work as the oob checks are also performed within the array before parsing the actual content.. But [then](https://github.com/bro/binpac/pull/4#issuecomment-336256505) a couple of Bro's parsers fail the tests. However, without the knowledge of how the affected protocols work it's tough to audit the parsers.

My idea is that there is/are some declarations in the formal specs of those parsers that use the wrong length for arrays and therefore without the `goto` that kind of save their ass they fail (the only thing I changed was to remove the `goto` from the for-loop that parsers known-length arrays)

jsiwek commented:
Yeah, thanks for the patience/explanations.  I've got more changes in the topic/jsiwek/bit-1829 branches of binpac and bro repos now.

https://github.com/bro/binpac/commit/471fa6fbabaec2e6709a8c38715adde8d693913d

So that binpac patch should have a similar result to your proposed patch (explained in 2nd bullet point of commit message) plus a couple other improvements.

https://github.com/bro/bro/commit/7b1421bd783e32b62653493028184f5d7296718b

And those were the fixes in Bro required for tests to pass -- it did appear that the MySQL parser was just generally broken and the SMB parser may have had a small typo or copy-paste-error that ended up relying on the faulty array parsing behavior (the old way would just ignore the field that was not supposed to be there in the first place).

What do you think of those changes?

freetom commented:
> 471fa6f
This looks good. However when I compile the sample parser I get this check pre-for-loop:
```
// Check bounds for static-size array: A:array
	if ( t_begin_of_data +  > t_end_of_data || t_begin_of_data +  < t_begin_of_data )
		throw binpac::ExceptionOutOfBound("A:array",
		  (), (t_end_of_data) - (t_begin_of_data));
```
It looks like the array size is gone lost somehow.

I have also tested the case in which the array is made of other objects of dynamic size and it works well (apart from the disappeared array size..)

[ExampleParser2.zip](https://github.com/bro/binpac/files/2014258/ExampleParser2.zip)
 

> bro/bro@7b1421b

about this idk actually

jsiwek commented:
That's baffling for several reasons.

I don't see the same thing when running that example through binpac: it does produce a static-size bounds check for B:array, but does not for A:array (which I think is correct).

So yours shouldn't have even generated the static-size check for A:array, and it's especially weird that the array_size string is empty -- that implies the fmt() function didn't even work...

Can you double-check what you did and maybe even provide me with explicit, step-by-step instructions from a fresh git checkout to reproduce this?

jsiwek commented:
Nevermind, seems I can reproduce on Linux and should be able to debug it.

jsiwek commented:
Should be fixed now if you're kind enough to give it one more try to confirm.

freetom commented:
```
// Check bounds for static-size array: B:array
	if ( t_begin_of_data + ((4) * (t_array__arraylength)) > t_end_of_data || t_begin_of_data + ((4) * (t_array__arraylength)) < t_begin_of_data )
		throw binpac::ExceptionOutOfBound("B:array",
		  (((4) * (t_array__arraylength))), (t_end_of_data) - (t_begin_of_data));
```
It's okay now.

jsiwek commented:
The fixes are now in master.
